File: BillPayments/Features/Bills/Data/DueBillsStore.swift
import Foundation
import Observation

@Observable
final class DueBillsStore {
    private(set) var dueBills: Async<[FetchedBill]> = .loading(previous: nil)
    @ObservationIgnored private(set) var pendingWork: Task<Void, Never>?
    @ObservationIgnored private(set) var observationWork: Task<Void, Never>?

    @ObservationIgnored private let savedBillersStore: SavedBillersStore
    @ObservationIgnored private let repository: BillRepository
    @ObservationIgnored private let freshness: Freshness

    init(
        authStore: AuthStore,
        savedBillersStore: SavedBillersStore,
        repository: BillRepository,
        clock: Clock
    ) {
        self.savedBillersStore = savedBillersStore
        self.repository = repository
        self.freshness = Freshness(clock: clock, maxAge: 300)
        observationWork = Task { [weak self, authStore, savedBillersStore] in
            let inputs = Observations {
                _ = authStore.state.userIdOrNull
                return savedBillersStore.state.items
            }
            for await items in inputs {
                guard let self else { return }
                self.fetch(items)
            }
        }
    }

    func refreshIfStale() {
        if case .loading = dueBills { return }
        guard freshness.isStale() else { return }
        invalidate()
    }

    func invalidate() {
        fetch(savedBillersStore.state.items)
    }

    private func fetch(_ saved: [SavedBiller]) {
        pendingWork?.cancel()
        if saved.isEmpty {
            dueBills = .data([])
            return
        }
        dueBills = .loading(previous: dueBills.valueOrNull)
        pendingWork = Task {
            do {
                let bills = try await repository.fetchDueBills(saved: saved)
                if Task.isCancelled { return }
                dueBills = .data(bills)
                freshness.markFetched()
            } catch {
                if Task.isCancelled { return }
                dueBills = .error(error)
            }
        }
    }
}

File: app/src/main/java/com/billpayments/features/bills/data/DueBillsStore.kt
package com.billpayments.features.bills.data

import com.billpayments.core.async.Async
import com.billpayments.core.cache.Freshness
import com.billpayments.core.time.Clock
import com.billpayments.features.auth.data.AuthStore
import com.billpayments.features.auth.data.userIdOrNull
import com.billpayments.features.saved_billers.data.SavedBiller
import com.billpayments.features.saved_billers.data.SavedBillersStore
import java.time.Duration
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch

@Singleton
class DueBillsStore @Inject constructor(
    authStore: AuthStore,
    private val savedBillersStore: SavedBillersStore,
    private val repository: BillRepository,
    clock: Clock,
    private val scope: CoroutineScope,
) {
    private val freshness = Freshness(clock, Duration.ofMinutes(5))
    private val _dueBills = MutableStateFlow<Async<List<FetchedBill>>>(Async.Loading)
    val dueBills: StateFlow<Async<List<FetchedBill>>> = _dueBills.asStateFlow()

    init {
        scope.launch {
            combine(
                authStore.state.map { it.userIdOrNull }.distinctUntilChanged(),
                savedBillersStore.state,
            ) { userId, saved -> if (userId == null) emptyList() else saved.items }
                .distinctUntilChanged()
                .collectLatest { fetch(it) }
        }
    }

    private suspend fun fetch(saved: List<SavedBiller>) {
        if (saved.isEmpty()) {
            _dueBills.value = Async.Data(emptyList())
            return
        }
        _dueBills.value = Async.Loading
        try {
            _dueBills.value = Async.Data(repository.fetchDueBills(saved))
            freshness.markFetched()
        } catch (e: Exception) {
            _dueBills.value = Async.Error(e)
        }
    }

    fun refreshIfStale() {
        if (_dueBills.value is Async.Loading || !freshness.isStale()) return
        invalidate()
    }

    fun invalidate() {
        scope.launch { fetch(savedBillersStore.state.value.items) }
    }
}
